### DISCUSSION TITLE
Asphere Questions

### DISCUSSION THREAD

**Discussion by cbhirsch** [mjhoptics/ray-optics#117] [Q&A] [answered]
So I'm looking to model Edmund Optics PN# 49665 with the below code:

```
sm.add_surface([63.00, 2.58, 'S-TIH13','OHARA', 11.5])
sm.ifcs[sm.cur_surface].profile = EvenPolynomial(r=63.00, cc=0,
	coefs=[0.0,0.0,4.662529E-6,-8.0842E-09,0.0,0.0,0.0])
sm.add_surface([31.0, 9.0, 'S-FSL5','OHARA', 11.25])
sm.add_surface([-28.5, 9.0, 'air', 11.25])
```
https://www.edmundoptics.com/p/25mm-diameter-x-50mm-efl-aspherized-achromatic-lens/10170/

I'm getting a bizarre output:

![image](https://user-images.githubusercontent.com/91820473/234403910-8810e3e6-6d34-4981-a589-32e87e3a4dc5.png)

Is there anything that I'm missing or doing wrong?

**Comment by mjhoptics** [comment]
Hello @cbhirsch,
I think I see the problem. I downloaded the CODE V and Zemax files from Edmund and it looks like you've flipped the lens around in your system. This means the local z axis of the surfaces is opposite to how the surfaces are defined by Edmund. To compensate for this, you've flipped the signs of the surface radii (good) but not the polynomial coefficients (bad). Flip the signs of the aspheric coefficients and I think that will fix your model.
HTH,
MIke

PS: There is a method in `OpticalModel` called [flip()](https://ray-optics.readthedocs.io/en/latest/api/rayoptics.optical.opticalmodel.html#rayoptics.optical.opticalmodel.OpticalModel.flip) that handles the details of flipping optical components such as this.

**Reply by mjhoptics** [reply, marked as answer]
One more thing - the aspheric coefficients are off by 1: the non-zero values should be 4th and 6th order coefs, instead they're entered as 6th and 8th order coefs. The coefficient list starts at the quadratic term. The correct entry should be
```
sm.ifcs[sm.cur_surface].profile = EvenPolynomial(r=63.00, cc=0,
	coefs=[0.0,-4.662529E-6,8.0842E-09,0.0,0.0,0.0])
```

**Reply by cbhirsch** [reply]
That fixed it so I'm guessing k doesn't show up then.  That's the mistake that I made.

**Reply by cbhirsch** [reply]
What does cc stand for? Would that be the k term in this example?

**Reply by mjhoptics** [reply]
`cc` is conic constant. CODE V calls it K. If cc=0, the base shape is a sphere.

Some of the literature (e.g. Spencer-Murty) use a eccentricity factor `ec` where ec=0 is a paraboloid. 

The input args for the aspheric profiles are set up to allow either `cc` or `ec` as input.

**Reply by cbhirsch** [reply]
Thanks for the clarification!
